core/platform/email/gae_email_services_test.py

```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+
+    def test_that_correct_bulk_emails_sent(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 1)
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 1)
+
+    def test_that_bulk_emails_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        email_exception = (self.assertRaisesRegexp(
+            Exception, 'This app cannot send emails.'))
+        with email_exception:
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.recipient_emails,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 0)
+
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 0)
+
+    def test_that_bulk_mails_not_sent_if_sender_email_is_malformed(self):
+
+        malformed_sender_email = ''
+        email_exception = (self.assertRaisesRegexp(
+            ValueError, 'Malformed sender email address: %s'
+            % malformed_sender_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_bulk_mail(
+                malformed_sender_email, self.recipient_emails,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(0, len(message_a))
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(0, len(message_b))
+
+    def test_that_bulk_mails_not_sent_if_recipient_email_is_malformed(self):
+        malformed_recipient_emails = ['', '']
+        email_exception = (self.assertRaisesRegexp(
+            ValueError, 'Malformed recipient email address: %s'
+            % malformed_recipient_emails[0]))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, malformed_recipient_emails,
+                'subject', 'body', 'html')
+        messages_a = self.mail_stub.get_sent_messages(
+            to=malformed_recipient_emails[0])
+        self.assertEqual(len(messages_a), 0)
+        messages_b = self.mail_stub.get_sent_messages(
+            to=malformed_recipient_emails[1])
+        self.assertEqual(len(messages_b), 0)
+        malformed_recipient_emails = [self.RECIPIENT_A_EMAIL, '']
```

Comment:
Add a newline above this, since it's basically a separate "chunk" of the test.

Reply:
Done